### DISCUSSION TITLE
formatting with null-ls

### DISCUSSION THREAD

**Discussion by ckruse** [LazyVim/LazyVim#100] [Q&A]
Hi there,

is it possible to disable autoformatting with jsonls and instead use prettierd?

I tried to achieve this with this config:

~~~ lua
return {
  {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function(_, opts)
      local nls = require("null-ls")
      return {
        nls.builtins.formatting.prettierd,
        nls.builtins.diagnostics.eslint_d,
        nls.builtins.diagnostics.ansiblelint,
      }
    end,
  },

  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        jsonls = {
          settings = {
            json = {
              format = {
                enable = false,
              },
            }
          }
        },
      },
    }
  }
}
~~~

but neovim still uses jsonls to format the json file.

**Edit:** it looks like formatting when modifying the null-ls config doesn't work at all. Given this:

```
return {
  {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function(_, opts)
      local nls = require("null-ls")
      opts.sources = {
        nls.builtins.formatting.stylua,
        nls.builtins.formatting.prettierd,
        nls.builtins.diagnostics.eslint_d,
        nls.builtins.diagnostics.ansiblelint,
      }

      return opts
    end,
  }
}
```

Now formatting on save seems to be completely disabled. But `prettierd` is installed:

<img width="277" alt="Bildschirm­foto 2023-01-23 um 12 57 43" src="https://user-images.githubusercontent.com/175095/214034254-36598eab-2e69-4d25-a29d-ee1e9718551d.png">

And `NullLsInfo` says, that it is enabled:

<img width="1083" alt="Bildschirm­foto 2023-01-23 um 12 58 35" src="https://user-images.githubusercontent.com/175095/214034532-6948aeb1-79d0-49f5-8396-efd5d1683023.png">

**Edit:** using `prettier` instead of `prettierd` works.

Best regards,
 CK

**Comment by AigenUK** [comment]
Hi @ckruse, did you ever get this working? I'm having similar issues and pulling my hair out! Trying to stop the formatting of double quotes to single and no semi-colons in .js
Any help would be appreciated.

**Comment by xulongwu4** [comment]
@ckruse, you can try this to disable jsonls's formatting function:

```
    opts = {
      servers = {
        jsonls = {
          -- Do not use jsonls's formatting capability. Use prettier to format json file
          init_options = { provideFormatter = false },
          settings = {jsonls = {...}}          
         },
      }
   }
```

**Reply by folke** [reply]
You don't need to disable it. LSP formatting is automatically disabled if a null-ls formatter exists for the buffer.

**Comment by smokeyhallow** [comment]
> Hi there,
> 
> is it possible to disable autoformatting with jsonls and instead use prettierd?
> 
> I tried to achieve this with this config:
> 
> ```lua
> return {
>   {
>     "jose-elias-alvarez/null-ls.nvim",
>     opts = function(_, opts)
>       local nls = require("null-ls")
>       return {
>         nls.builtins.formatting.prettierd,
>         nls.builtins.diagnostics.eslint_d,
>         nls.builtins.diagnostics.ansiblelint,
>       }
>     end,
>   },
> 
>   {
>     "neovim/nvim-lspconfig",
>     opts = {
>       servers = {
>         jsonls = {
>           settings = {
>             json = {
>               format = {
>                 enable = false,
>               },
>             }
>           }
>         },
>       },
>     }
>   }
> }
> ```
> 
> but neovim still uses jsonls to format the json file.
> 
> **Edit:** it looks like formatting when modifying the null-ls config doesn't work at all. Given this:
> 
> ```
> return {
>   {
>     "jose-elias-alvarez/null-ls.nvim",
>     opts = function(_, opts)
>       local nls = require("null-ls")
>       opts.sources = {
>         nls.builtins.formatting.stylua,
>         nls.builtins.formatting.prettierd,
>         nls.builtins.diagnostics.eslint_d,
>         nls.builtins.diagnostics.ansiblelint,
>       }
> 
>       return opts
>     end,
>   }
> }
> ```
> 
> Now formatting on save seems to be completely disabled. But `prettierd` is installed:
> 
> <img alt="Bildschirm­foto 2023-01-23 um 12 57 43" width="277" src="https://user-images.githubusercontent.com/175095/214034254-36598eab-2e69-4d25-a29d-ee1e9718551d.png">
> 
> And `NullLsInfo` says, that it is enabled:
> 
> <img alt="Bildschirm­foto 2023-01-23 um 12 58 35" width="1083" src="https://user-images.githubusercontent.com/175095/214034532-6948aeb1-79d0-49f5-8396-efd5d1683023.png">
> 
> **Edit:** using `prettier` instead of `prettierd` works.
> 
> Best regards, CK

Exactly my issue! I can even see `prettierd` in process list. There is even notification on save that prettier is running but its just not getting formatted.

**Reply by folke** [reply]
It works perfectly fine for me, so something in your config must be wrong.

There's an extra for setting up prettierd with null-ls. Check the source to see how to set it up
